In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
1,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
2,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
3,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
4,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# LASSO (L1-regularized Logistic Regression)

## 2. Baseline

In [10]:
# =============================================================================
# LASSO BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from sklearn.linear_model import LogisticRegression
import pandas as pd
import numpy as np


# Fit LASSO once with all features
# Logistic Regression with L1 penalty = LASSO-style feature selection

lasso_selector = LogisticRegression(
    penalty="l1",
    solver="liblinear",
    C=0.1,
    max_iter=2000,
    class_weight="balanced",
    random_state=42
)


lasso_selector.fit(
    X_train_scaled,
    y_train
)


# =============================================================================
# Extract LASSO coefficients
# =============================================================================

lasso_coefficients = lasso_selector.coef_[0]


# Use absolute coefficient magnitude as feature importance
lasso_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "LASSO Coefficient": lasso_coefficients,

    "LASSO Score": np.abs(lasso_coefficients)

})


# Sort by absolute coefficient magnitude
lasso_scores = lasso_scores.sort_values(
    by="LASSO Score",
    ascending=False
).reset_index(drop=True)


# =============================================================================
# Top-K values
# =============================================================================

k_values = [5, 10, 15, 20]

lasso_results = {}


for k in k_values:

    top_features = (
        lasso_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    lasso_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": lasso_scores.head(k)

    }


    print("=" * 70)
    print(f"TOP {k} LASSO FEATURES")
    print("=" * 70)

    print(top_features)

    print("\nScores:")
    print(lasso_scores.head(k))

TOP 5 LASSO FEATURES
['worst radius', 'worst concave points', 'worst texture', 'radius error', 'mean concave points']

Scores:
                Feature  LASSO Coefficient  LASSO Score
0          worst radius          -2.079876     2.079876
1  worst concave points          -1.084678     1.084678
2         worst texture          -0.626262     0.626262
3          radius error          -0.517377     0.517377
4   mean concave points          -0.405840     0.405840
TOP 10 LASSO FEATURES
['worst radius', 'worst concave points', 'worst texture', 'radius error', 'mean concave points', 'worst symmetry', 'worst smoothness', 'mean texture', 'mean symmetry', 'mean smoothness']

Scores:
                Feature  LASSO Coefficient  LASSO Score
0          worst radius          -2.079876     2.079876
1  worst concave points          -1.084678     1.084678
2         worst texture          -0.626262     0.626262
3          radius error          -0.517377     0.517377
4   mean concave points          -0.405

In [11]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [12]:
# =============================================================================
# MODEL EVALUATION — LASSO
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results_lasso = []


for k in k_values:

    Xtr = lasso_results[k]["X_train"]

    Xts = lasso_results[k]["X_test"]

    features = ", ".join(
        lasso_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        results_lasso.append({

            "Method": "LASSO",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred
            ),

            "Recall": recall_score(
                y_test,
                y_pred
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


lasso_evaluation = pd.DataFrame(
    results_lasso
)

display(lasso_evaluation)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO,5,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.993386
1,LASSO,5,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.995040
2,LASSO,5,XGBoost,"worst radius, worst concave points, worst text...",0.947368,0.958333,0.958333,0.958333,0.994048
3,LASSO,10,Logistic Regression,"worst radius, worst concave points, worst text...",0.973684,0.985915,0.972222,0.979021,0.996032
4,LASSO,10,Random Forest,"worst radius, worst concave points, worst text...",0.964912,0.985714,0.958333,0.971831,0.994709
5,LASSO,10,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.958904,0.972222,0.965517,0.992394
6,LASSO,15,Logistic Regression,"worst radius, worst concave points, worst text...",0.947368,0.985294,0.930556,0.957143,0.993717
7,LASSO,15,Random Forest,"worst radius, worst concave points, worst text...",0.947368,0.971429,0.944444,0.957746,0.994709
8,LASSO,15,XGBoost,"worst radius, worst concave points, worst text...",0.964912,0.972222,0.972222,0.972222,0.995040
9,LASSO,20,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.985507,0.944444,0.964539,0.994709


In [13]:
lasso_evaluation.to_csv(
    "../../../results/breast_cancer/baseline/lasso_baseline_results.csv",
    index=False
)

## 2. DODA

In [14]:
%pip uninstall -y doda

Found existing installation: doda 0.1.0
Uninstalling doda-0.1.0:
  Successfully uninstalled doda-0.1.0
Note: you may need to restart the kernel to use updated packages.


In [15]:
%pip install git+https://github.com/anandha-3679/DODA.git

  Cloning https://github.com/anandha-3679/DODA.git to C:\Users\johnm\AppData\Local\Temp\pip-req-build-af_id7pu
  Resolved https://github.com/anandha-3679/DODA.git to commit ff8eb84b57e3c9c01f5028f902c0fd76be092333
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for doda: filename=doda-0.1.0-py3-none-any.whl size=13504 sha256=13706049eaee42cff7ae3b264ef86a661c8f4ad71ac1ccab7a3b72ec0b9a0360
  Stored in directory: C:\Users\johnm\AppData\Local\Temp\pip-ephem-wheel-cache-t166d0p1\wheels\42\6a\22\a51a03690cc18d3493705062872e3519eede94dc00682c19bc
Successfully built doda
Note: you may need to restart the kernel to use updated packages.


  Running command git clone --filter=blob:none --quiet https://github.com/anandha-3679/DODA.git 'C:\Users\johnm\AppData\Local\Temp\pip-req-build-af_id7pu'


In [16]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [17]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [18]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [19]:
# =============================================================================
# LASSO + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.linear_model import LogisticRegression


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN LASSO + DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"LASSO + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # LASSO operator
    # ---------------------------------------------------------------------

    lasso = SklearnAdapter(

        LogisticRegression(
            penalty="l1",
            solver="liblinear",
            C=0.1,
            max_iter=2000,
            class_weight="balanced",
            random_state=42
        )

    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(

        operators=[lasso],

        provider=provider,

        fusion=fusion,

        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(

        X_train_scaled,

        y_train

    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(

        X_test_scaled

    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = (
        selector.get_selected_features()
    )


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores":
            selector.raw_math_scores_,

        "math_scores":
            selector.math_scores_,

        "clinical_weights":
            selector.clinical_weights_,

        "final_scores":
            selector.final_scores_
    }


print("\n" + "=" * 70)
print("LASSO + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


LASSO + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x0000025B292C9010>
DODA Pipeline Started
Running: LogisticRegression
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of coefficients: 30
Operator Scores:
{'LogisticRegression_1': {'mean radius': np.float64(0.0), 'mean texture': np.float64(0.03116243353789412), 'mean perimeter': np.float64(0.0), 'mean area': np.float64(0.0), 'mean smoothness': np.float64(0.0), 'mean compactness': np.float64(0.0), 'mean concavity': np.float64(0.0), 'mean concave points': np.float64(0.4058401456180695), 'mean symmetry': np.float64(0.0), 'mean fractal dimension': np.float64(0.0), 'radius error': np.float64(0.5173770695234363), 'texture error': np.float64(0.0), 'perimeter error': np.float64(0.0),

In [20]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
20,worst radius,2.079876,1.000000
27,worst concave points,1.084678,0.521511
21,worst texture,0.626262,0.301106
10,radius error,0.517377,0.248754
7,mean concave points,0.405840,0.195127
28,worst symmetry,0.152786,0.073459
24,worst smoothness,0.151605,0.072891
1,mean texture,0.031162,0.014983
8,mean symmetry,0.000000,0.000000
4,mean smoothness,0.000000,0.000000


In [21]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
27,worst concave points,1.0
26,worst concavity,1.0
22,worst perimeter,1.0
23,worst area,1.0
20,worst radius,1.0
0,mean radius,0.9
3,mean area,0.9
6,mean concavity,0.9
7,mean concave points,0.9
2,mean perimeter,0.9


In [22]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
20,worst radius,1.000000
27,worst concave points,0.521511
21,worst texture,0.180663
7,mean concave points,0.175614
10,radius error,0.124377
28,worst symmetry,0.044075
24,worst smoothness,0.036446
1,mean texture,0.007491
8,mean symmetry,0.000000
4,mean smoothness,0.000000


In [23]:
selector.get_selected_features()

['worst radius',
 'worst concave points',
 'worst texture',
 'mean concave points',
 'radius error',
 'worst symmetry',
 'worst smoothness',
 'mean texture',
 'mean radius',
 'mean perimeter']

In [24]:
comparison = selector.compare_scores()

comparison

,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,worst radius,1,1.0000,1.0,1,1.0000,0
1,worst concave points,2,0.5215,1.0,2,0.5215,0
2,worst texture,3,0.3011,0.6,3,0.1807,0
3,mean concave points,5,0.1951,0.9,4,0.1756,1
4,radius error,4,0.2488,0.5,5,0.1244,-1
5,worst symmetry,6,0.0735,0.6,6,0.0441,0
6,worst smoothness,7,0.0729,0.5,7,0.0364,0
7,mean texture,8,0.0150,0.5,8,0.0075,0
8,mean radius,9,0.0000,0.9,9,0.0000,0
9,mean perimeter,10,0.0000,0.9,10,0.0000,0


In [25]:
# =============================================================================
# LASSO + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


# -------------------------------------------------------------------------
# Store results
# -------------------------------------------------------------------------

lasso_doda_model_results = []


# -------------------------------------------------------------------------
# Evaluate for each Top-K
# -------------------------------------------------------------------------

for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    # ---------------------------------------------------------------------
    # Evaluate each model
    # ---------------------------------------------------------------------

    for model_name, model in models.items():

        # Train model
        model.fit(
            Xtr,
            y_train
        )


        # Predictions
        y_pred = model.predict(
            Xts
        )


        # Probability predictions
        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # ---------------------------------------------------------------
        # Store evaluation metrics
        # ---------------------------------------------------------------

        lasso_doda_model_results.append({

            "Method": "LASSO + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# -------------------------------------------------------------------------
# Convert results to DataFrame
# -------------------------------------------------------------------------

lasso_doda_results_df = pd.DataFrame(
    lasso_doda_model_results
)


# -------------------------------------------------------------------------
# Display results
# -------------------------------------------------------------------------

display(
    lasso_doda_results_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO + DODA,5,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.993386
1,LASSO + DODA,5,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.958904,0.972222,0.965517,0.995040
2,LASSO + DODA,5,XGBoost,"worst radius, worst concave points, worst text...",0.964912,0.972222,0.972222,0.972222,0.993717
3,LASSO + DODA,10,Logistic Regression,"worst radius, worst concave points, worst text...",0.938596,0.985075,0.916667,0.949640,0.994709
4,LASSO + DODA,10,Random Forest,"worst radius, worst concave points, worst text...",0.938596,0.971014,0.930556,0.950355,0.994709
5,LASSO + DODA,10,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.994709
6,LASSO + DODA,15,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.985507,0.944444,0.964539,0.993717
7,LASSO + DODA,15,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.994213
8,LASSO + DODA,15,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.994378
9,LASSO + DODA,20,Logistic Regression,"worst radius, worst concave points, worst text...",0.947368,0.985294,0.930556,0.957143,0.994709


In [26]:
lasso_doda_results_df.to_csv(
    "../../../results/breast_cancer/baseline/lasso_doda_results.csv",
    index=False
)

In [27]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        lasso_evaluation,
        lasso_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO,5,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.993386
1,LASSO,5,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.995040
2,LASSO,5,XGBoost,"worst radius, worst concave points, worst text...",0.947368,0.958333,0.958333,0.958333,0.994048
3,LASSO,10,Logistic Regression,"worst radius, worst concave points, worst text...",0.973684,0.985915,0.972222,0.979021,0.996032
4,LASSO,10,Random Forest,"worst radius, worst concave points, worst text...",0.964912,0.985714,0.958333,0.971831,0.994709
5,LASSO,10,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.958904,0.972222,0.965517,0.992394
6,LASSO,15,Logistic Regression,"worst radius, worst concave points, worst text...",0.947368,0.985294,0.930556,0.957143,0.993717
7,LASSO,15,Random Forest,"worst radius, worst concave points, worst text...",0.947368,0.971429,0.944444,0.957746,0.994709
8,LASSO,15,XGBoost,"worst radius, worst concave points, worst text...",0.964912,0.972222,0.972222,0.972222,0.995040
9,LASSO,20,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.985507,0.944444,0.964539,0.994709


In [28]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['worst radius', 'worst concave points', 'worst texture', 'radius error', 'mean concave points', 'worst symmetry', 'worst smoothness', 'mean texture', 'mean radius', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean symmetry', 'mean fractal dimension', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst perimeter', 'worst area', 'worst compactness', 'worst concavity', 'worst fractal dimension']

DODA Ranking:
['worst radius', 'worst concave points', 'worst texture', 'mean concave points', 'radius error', 'worst symmetry', 'worst smoothness', 'mean texture', 'mean radius', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean symmetry', 'mean fractal dimension', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness erro

# 3. LASSO + DODA: Model Evaluation & Interpretation

---

## 3.1 Overall Predictive Performance

### 📊 Performance by Top-K

#### Top-K = 5

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9561 → 0.9561 | 0.9718 → 0.9718 | 0.9583 → 0.9583 | 0.9650 → 0.9650 | 0.9934 → 0.9934 | 🟡 Maintained |
| Random Forest | 0.9561 → 0.9561 | 0.9718 → 0.9589 | 0.9583 → 0.9722 | 0.9650 → 0.9655 | 0.9950 → 0.9950 | 🟡 Comparable |
| XGBoost | 0.9474 → **0.9649** | 0.9583 → **0.9722** | 0.9583 → **0.9722** | 0.9583 → **0.9722** | 0.9940 → 0.9937 | 🟢 Improved |

#### Top-K = 10

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9737 → 0.9386 | 0.9859 → 0.9851 | 0.9722 → 0.9167 | 0.9790 → 0.9496 | 0.9960 → 0.9947 | 🔴 Reduced |
| Random Forest | 0.9649 → 0.9386 | 0.9857 → 0.9710 | 0.9583 → 0.9306 | 0.9718 → 0.9504 | 0.9947 → 0.9947 | 🔴 Reduced |
| XGBoost | 0.9561 → 0.9561 | 0.9589 → 0.9718 | 0.9722 → 0.9583 | 0.9655 → 0.9650 | 0.9924 → 0.9947 | 🟡 Comparable |

#### Top-K = 15

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9474 → **0.9561** | 0.9853 → 0.9855 | 0.9306 → **0.9444** | 0.9571 → **0.9645** | 0.9937 → 0.9937 | 🟢 Improved |
| Random Forest | 0.9474 → **0.9561** | 0.9714 → 0.9718 | 0.9444 → **0.9583** | 0.9577 → **0.9650** | 0.9947 → 0.9942 | 🟢 Improved |
| XGBoost | 0.9649 → 0.9561 | 0.9722 → 0.9718 | 0.9722 → 0.9583 | 0.9722 → 0.9650 | 0.9950 → 0.9944 | 🔴 Reduced |

#### Top-K = 20

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9561 → 0.9474 | 0.9855 → 0.9853 | 0.9444 → 0.9306 | 0.9645 → 0.9571 | 0.9947 → 0.9947 | 🔴 Reduced |
| Random Forest | 0.9561 → 0.9474 | 0.9718 → 0.9583 | 0.9583 → 0.9583 | 0.9650 → 0.9583 | 0.9940 → 0.9937 | 🔴 Reduced |
| XGBoost | 0.9737 → 0.9474 | 0.9859 → 0.9459 | 0.9722 → 0.9722 | 0.9790 → 0.9589 | 0.9967 → 0.9964 | 🔴 Reduced |

> **Notation:** `LASSO → DODA`  
> 🟢 Improved &nbsp;&nbsp; 🟡 Maintained / comparable &nbsp;&nbsp; 🔴 Reduced

---

### 🔎 Overall Performance Pattern

| DODA Effect | Configurations |
|:------------|:---------------|
| 🟢 **Improved** | XGBoost Top-K 5; Logistic Regression & Random Forest Top-K 15 |
| 🟡 **Maintained / Comparable** | Logistic Regression & Random Forest Top-K 5; XGBoost Top-K 10 |
| 🔴 **Reduced** | Logistic Regression & Random Forest Top-K 10; XGBoost Top-K 15; all models Top-K 20 |

**Key finding:** DODA does not consistently improve predictive performance. Its effect depends on the model and Top-K feature size.

The strongest LASSO baseline results were:

- **Logistic Regression, Top-K = 10:** Accuracy = **0.9737**, F1 = **0.9790**
- **XGBoost, Top-K = 20:** Accuracy = **0.9737**, F1 = **0.9790**, ROC-AUC = **0.9967**

The strongest DODA result by accuracy and F1-score was:

- **XGBoost, Top-K = 5:** Accuracy = **0.9649**, F1 = **0.9722**

---

## 3.2 Effect of DODA on Feature Selection

### 📊 Top-10 Ranking Comparison

| Rank | LASSO | DODA | Change |
|-----:|:------|:-----|:------|
| 1 | `worst radius` | `worst radius` | — |
| 2 | `worst concave points` | `worst concave points` | — |
| 3 | `worst texture` | `worst texture` | — |
| 4 | `radius error` | `mean concave points` | ⬆️ |
| 5 | `mean concave points` | `radius error` | ⬇️ |
| 6 | `worst symmetry` | `worst symmetry` | — |
| 7 | `worst smoothness` | `worst smoothness` | — |
| 8 | `mean texture` | `mean texture` | — |
| 9 | `mean symmetry` | `mean radius` | ↕️ |
| 10 | `mean smoothness` | `mean perimeter` | ↕️ |

### 🎯 Main Feature Changes

- **Top-5 feature set:** unchanged, but internal ranking changed.
- `mean concave points`: **Rank 5 → 4** ⬆️
- `radius error`: **Rank 4 → 5** ⬇️
- `mean radius` and `mean perimeter` entered the DODA Top-10.
- `mean concavity` entered the DODA Top-15.
- `smoothness error` entered the DODA Top-20.
- Several later changes occurred among features with zero mathematical and DODA scores and should therefore be interpreted cautiously.

---

## 3.3 Clinical Weight Alignment

| Feature | LASSO Rank | DODA Rank | Clinical Weight | Direction |
|:--------|-----------:|----------:|----------------:|:---------:|
| `mean concave points` | 5 | **4** | **0.9** | ⬆️ |
| `radius error` | 4 | 5 | 0.5 | ⬇️ |
| `mean radius` | 14 | **9** | **0.9** | ⬆️ |
| `mean perimeter` | 13 | **10** | **0.9** | ⬆️ |
| `mean concavity` | 19 | **14** | **0.9** | ⬆️ |
| `mean symmetry` | 9 | 15 | 0.5 | ⬇️ |
| `mean smoothness` | 10 | 12 | 0.4 | ⬇️ |

### 🩺 Interpretation

The clearest clinically aligned scored change was:

```text
radius error          → Rank 4 → 5
mean concave points   → Rank 5 → 4